File: objc/Tests/TestTransportConformance.m
/*
 * TestTransportConformance — v0.10 M70.
 *
 * In-language .tio → .tis → .tio round-trip with signal values
 * preserved bit-for-bit.
 *
 * SPDX-License-Identifier: Apache-2.0
 */
#import <Foundation/Foundation.h>
#import "Testing.h"
#import <unistd.h>
#include <math.h>

#import "Transport/TTIOTransportWriter.h"
#import "Transport/TTIOTransportReader.h"
#import "Dataset/TTIOSpectralDataset.h"
#import "Dataset/TTIOWrittenRun.h"
#import "Run/TTIOAcquisitionRun.h"
#import "Spectra/TTIOSpectrum.h"
#import "Spectra/TTIOMassSpectrum.h"
#import "Core/TTIOSignalArray.h"
#import "ValueClasses/TTIOEnums.h"

static NSString *tmp(NSString *n) {
    return [NSString stringWithFormat:@"/tmp/ttio_m70_%d_%@", (int)getpid(), n];
}
static void rm(NSString *p) { [[NSFileManager defaultManager] removeItemAtPath:p error:NULL]; }

static NSData *f64le(const double *v, NSUInteger n) {
    NSMutableData *d = [NSMutableData dataWithCapacity:n * 8];
    for (NSUInteger i = 0; i < n; i++) [d appendBytes:&v[i] length:8];
    return d;
}
static NSData *i32arr(const int32_t *v, NSUInteger n) {
    NSMutableData *d = [NSMutableData dataWithCapacity:n * 4];
    for (NSUInteger i = 0; i < n; i++) [d appendBytes:&v[i] length:4];
    return d;
}
static NSData *u32arr(const uint32_t *v, NSUInteger n) {
    NSMutableData *d = [NSMutableData dataWithCapacity:n * 4];
    for (NSUInteger i = 0; i < n; i++) [d appendBytes:&v[i] length:4];
    return d;
}
static NSData *u64arr(const uint64_t *v, NSUInteger n) {
    NSMutableData *d = [NSMutableData dataWithCapacity:n * 8];
    for (NSUInteger i = 0; i < n; i++) [d appendBytes:&v[i] length:8];
    return d;
}

static BOOL buildDataset(NSString *path, NSUInteger nRuns,
                           NSUInteger nSpectra, NSUInteger pointsPerSpectrum,
                           NSError **error)
{
    NSMutableDictionary<NSString *, TTIOWrittenRun *> *runMap = [NSMutableDictionary dictionary];
    for (NSUInteger r = 0; r < nRuns; r++) {
        NSUInteger total = nSpectra * pointsPerSpectrum;
        double *mz = calloc(total, sizeof(double));
        double *intensity = calloc(total, sizeof(double));
        for (NSUInteger i = 0; i < total; i++) {
            mz[i] = 100.0 * (r + 1) + (double)i;
            intensity[i] = 100.0 * (r + 1) * (double)(i + 1);
        }
        uint64_t *offsets = calloc(nSpectra, sizeof(uint64_t));
        uint32_t *lengths = calloc(nSpectra, sizeof(uint32_t));
        for (NSUInteger i = 0; i < nSpectra; i++) {
            offsets[i] = (uint64_t)(i * pointsPerSpectrum);
            lengths[i] = (uint32_t)pointsPerSpectrum;
        }
        double *rts = calloc(nSpectra, sizeof(double));
        int32_t *msLevels = calloc(nSpectra, sizeof(int32_t));
        int32_t *pols = calloc(nSpectra, sizeof(int32_t));
        double *pmzs = calloc(nSpectra, sizeof(double));
        int32_t *pcs = calloc(nSpectra, sizeof(int32_t));
        double *bpis = calloc(nSpectra, sizeof(double));
        for (NSUInteger i = 0; i < nSpectra; i++) {
            rts[i] = 1.0 + (double)i;
            msLevels[i] = (i % 2 == 0) ? 1 : 2;
            pols[i] = 1;
            pmzs[i] = msLevels[i] == 1 ? 0.0 : 500.0 + (double)i;
            pcs[i] = msLevels[i] == 1 ? 0 : 2;
            double best = 0.0;
            for (NSUInteger k = 0; k < pointsPerSpectrum; k++) {
                double v = intensity[i * pointsPerSpectrum + k];
                if (v > best) best = v;
            }
            bpis[i] = best;
        }
        TTIOWrittenRun *run =
            [[TTIOWrittenRun alloc]
                initWithSpectrumClassName:@"TTIOMassSpectrum"
                          acquisitionMode:(int64_t)TTIOAcquisitionModeMS1DDA
                              channelData:@{@"mz": f64le(mz, total),
                                            @"intensity": f64le(intensity, total)}
                                  offsets:u64arr(offsets, nSpectra)
                                  lengths:u32arr(lengths, nSpectra)
                           retentionTimes:f64le(rts, nSpectra)
                                 msLevels:i32arr(msLevels, nSpectra)
                               polarities:i32arr(pols, nSpectra)
                             precursorMzs:f64le(pmzs, nSpectra)
                         precursorCharges:i32arr(pcs, nSpectra)
                      basePeakIntensities:f64le(bpis, nSpectra)];
        runMap[[NSString stringWithFormat:@"run_%04lu", (unsigned long)r]] = run;
        free(mz); free(intensity);
        free(offsets); free(lengths); free(rts);
        free(msLevels); free(pols); free(pmzs); free(pcs); free(bpis);
    }
    return [TTIOSpectralDataset writeMinimalToPath:path
                                              title:@"M70 ObjC conformance"
                                 isaInvestigationId:@"ISA-M70-OBJC"
                                             msRuns:runMap
                                    identifications:nil
                                    quantifications:nil
                                  provenanceRecords:nil
                                              error:error];
}

static BOOL signalBytesEqual(NSData *a, NSData *b)
{
    if (a.length != b.length) return NO;
    return memcmp(a.bytes, b.bytes, a.length) == 0;
}

static BOOL datasetsSignalEqual(TTIOSpectralDataset *a, TTIOSpectralDataset *b)
{
    if (![[NSSet setWithArray:a.msRuns.allKeys]
            isEqualToSet:[NSSet setWithArray:b.msRuns.allKeys]]) return NO;
    for (NSString *name in a.msRuns) {
        TTIOAcquisitionRun *ra = a.msRuns[name];
        TTIOAcquisitionRun *rb = b.msRuns[name];
        if ([ra count] != [rb count]) return NO;
        for (NSUInteger i = 0; i < [ra count]; i++) {
            TTIOSpectrum *sa = [ra objectAtIndex:i];
            TTIOSpectrum *sb = [rb objectAtIndex:i];
            if (fabs(sa.scanTimeSeconds - sb.scanTimeSeconds) > 1e-12) return NO;
            if (fabs(sa.precursorMz - sb.precursorMz) > 1e-12) return NO;
            if (!signalBytesEqual(sa.signalArrays[@"mz"].buffer,
                                   sb.signalArrays[@"mz"].buffer)) return NO;
            if (!signalBytesEqual(sa.signalArrays[@"intensity"].buffer,
                                   sb.signalArrays[@"intensity"].buffer)) return NO;
        }
    }
    return YES;
}

static BOOL roundTrip(NSUInteger nRuns, NSUInteger nSpectra,
                        NSUInteger pointsPerSpectrum, BOOL withChecksum)
{
    NSString *src = tmp(@"src.tio");
    NSString *mots = tmp(@"stream.tis");
    NSString *rt = tmp(@"rt.tio");
    rm(src); rm(mots); rm(rt);

    NSError *err = nil;
    if (!buildDataset(src, nRuns, nSpectra, pointsPerSpectrum, &err)) return NO;

    TTIOSpectralDataset *source = [TTIOSpectralDataset readFromFilePath:src error:&err];
    TTIOTransportWriter *tw = [[TTIOTransportWriter alloc] initWithOutputPath:mots];
    tw.useChecksum = withChecksum;
    if (![tw writeDataset:source error:&err]) { [tw close]; return NO; }
    [tw close];

    TTIOTransportReader *tr = [[TTIOTransportReader alloc] initWithInputPath:mots];
    if (![tr writeTtioToPath:rt error:&err]) return NO;

    TTIOSpectralDataset *rtDs = [TTIOSpectralDataset readFromFilePath:rt error:&err];
    BOOL eq = datasetsSignalEqual(source, rtDs);
    rm(src); rm(mots); rm(rt);
    return eq;
}

static BOOL roundTripWithCompression(NSUInteger nRuns, NSUInteger nSpectra,
                                       NSUInteger pointsPerSpectrum)
{
    NSString *src = tmp(@"src-zlib.tio");
    NSString *mots = tmp(@"stream-zlib.tis");
    NSString *rt = tmp(@"rt-zlib.tio");
    rm(src); rm(mots); rm(rt);

    NSError *err = nil;
    if (!buildDataset(src, nRuns, nSpectra, pointsPerSpectrum, &err)) return NO;

    TTIOSpectralDataset *source = [TTIOSpectralDataset readFromFilePath:src error:&err];
    TTIOTransportWriter *tw = [[TTIOTransportWriter alloc] initWithOutputPath:mots];
    tw.useCompression = YES;
    if (![tw writeDataset:source error:&err]) { [tw close]; return NO; }
    [tw close];

    TTIOTransportReader *tr = [[TTIOTransportReader alloc] initWithInputPath:mots];
    if (![tr writeTtioToPath:rt error:&err]) return NO;

    TTIOSpectralDataset *rtDs = [TTIOSpectralDataset readFromFilePath:rt error:&err];
    BOOL eq = datasetsSignalEqual(source, rtDs);
    rm(src); rm(mots); rm(rt);
    return eq;
}

void testTransportConformance(void)
{
    PASS(roundTrip(1, 5, 4, NO), "single run, 5 spectra × 4 pts: round-trip bit-equal");
    PASS(roundTrip(3, 4, 5, NO), "3 runs, 4 spectra × 5 pts: round-trip bit-equal");
    PASS(roundTrip(1, 20, 128, NO), "1 run × 20 spectra × 128 pts: round-trip bit-equal");
    PASS(roundTrip(1, 5, 4, YES), "round-trip with CRC-32C checksum: bit-equal");
    PASS(roundTripWithCompression(1, 5, 4),
         "round-trip with ZLIB wire compression: bit-equal");
    PASS(roundTripWithCompression(1, 20, 128),
         "round-trip with ZLIB wire compression + larger spectra: bit-equal");
}

File: python/tests/test_transport_conformance.py
"""v0.10 M70: bidirectional conversion conformance.

Two layers of testing:

1. **In-language round-trip** — .tio → .tis → .tio with signal
   values preserved to float64 epsilon. Same for multi-run,
   multi-spectrum, and empty-run edge cases.

2. **Cross-language exchange** — Python spawns Java and ObjC
   encode/decode CLIs and verifies that a stream produced in one
   language is decodable in the other two.

Cross-language tests skip automatically when the Java classpath or
ObjC binary cannot be located (e.g. running outside the TTI-O
repo layout).
"""
from __future__ import annotations

import os
import shutil
import subprocess
from pathlib import Path

import numpy as np
import pytest

from ttio.enums import AcquisitionMode, Polarity
from ttio.spectral_dataset import SpectralDataset, WrittenRun
from ttio.transport.codec import file_to_transport, transport_to_file


REPO_ROOT = Path(__file__).resolve().parents[2]


# ---------------------------------------------------------- fixtures


def _build_dataset(path: Path, *, n_runs: int = 1, n_spectra: int = 5,
                    points_per_spectrum: int = 4) -> Path:
    runs: dict[str, WrittenRun] = {}
    for r in range(n_runs):
        total = n_spectra * points_per_spectrum
        mz = np.arange(total, dtype="<f8") + 100.0 * (r + 1)
        intensity = (np.arange(total, dtype="<f8") + 1.0) * (100.0 * (r + 1))
        offsets = np.arange(0, total, points_per_spectrum, dtype="<u8")
        lengths = np.full(n_spectra, points_per_spectrum, dtype="<u4")
        rts = np.linspace(1.0, float(n_spectra), n_spectra, dtype="<f8")
        ms_levels = np.array(
            [1 if i % 2 == 0 else 2 for i in range(n_spectra)], dtype="<i4"
        )
        polarities = np.full(n_spectra, int(Polarity.POSITIVE), dtype="<i4")
        precursor_mzs = np.array(
            [0.0 if ms_levels[i] == 1 else 500.0 + i for i in range(n_spectra)],
            dtype="<f8",
        )
        precursor_charges = np.array(
            [0 if ms_levels[i] == 1 else 2 for i in range(n_spectra)], dtype="<i4"
        )
        base_peaks = np.array([
            float(intensity[i * points_per_spectrum:(i + 1) * points_per_spectrum].max())
            for i in range(n_spectra)
        ], dtype="<f8")
        runs[f"run_{r:04d}"] = WrittenRun(
            spectrum_class="TTIOMassSpectrum",
            acquisition_mode=int(AcquisitionMode.MS1_DDA),
            channel_data={"mz": mz, "intensity": intensity},
            offsets=offsets,
            lengths=lengths,
            retention_times=rts,
            ms_levels=ms_levels,
            polarities=polarities,
            precursor_mzs=precursor_mzs,
            precursor_charges=precursor_charges,
            base_peak_intensities=base_peaks,
        )
    SpectralDataset.write_minimal(
        path,
        title="M70 conformance fixture",
        isa_investigation_id="ISA-M70",
        runs=runs,
    )
    return path


def _assert_signal_equal(a: SpectralDataset, b: SpectralDataset) -> None:
    assert set(a.all_runs) == set(b.all_runs)
    for name in a.all_runs:
        ra, rb = a.all_runs[name], b.all_runs[name]
        assert len(ra) == len(rb), f"run {name} spectrum count mismatch"
        for i in range(len(ra)):
            sa, sb = ra[i], rb[i]
            for c in ra.channel_names:
                aa = np.asarray(sa.signal_array(c).data)
                bb = np.asarray(sb.signal_array(c).data)
                np.testing.assert_allclose(
                    aa, bb, rtol=0.0, atol=0.0,
                    err_msg=f"run={name} spectrum={i} channel={c}"
                )
            assert sa.scan_time_seconds == pytest.approx(sb.scan_time_seconds)
            assert sa.precursor_mz == pytest.approx(sb.precursor_mz)


# ---------------------------------------------------------- in-language


class TestInLanguageRoundTrip:

    def test_single_run_roundtrip(self, tmp_path):
        src = _build_dataset(tmp_path / "src.tio")
        mots = tmp_path / "stream.tis"
        rt = tmp_path / "rt.tio"
        file_to_transport(src, mots)
        ds = transport_to_file(mots, rt)
        try:
            with SpectralDataset.open(src) as original:
                _assert_signal_equal(original, ds)
        finally:
            ds.close()

    def test_multi_run_roundtrip(self, tmp_path):
        src = _build_dataset(tmp_path / "src.tio", n_runs=3)
        mots = tmp_path / "stream.tis"
        rt = tmp_path / "rt.tio"
        file_to_transport(src, mots)
        ds = transport_to_file(mots, rt)
        try:
            with SpectralDataset.open(src) as original:
                _assert_signal_equal(original, ds)
        finally:
            ds.close()

    def test_larger_spectra(self, tmp_path):
        src = _build_dataset(
            tmp_path / "src.tio", n_spectra=20, points_per_spectrum=128
        )
        mots = tmp_path / "stream.tis"
        rt = tmp_path / "rt.tio"
        file_to_transport(src, mots)
        ds = transport_to_file(mots, rt)
        try:
            with SpectralDataset.open(src) as original:
                _assert_signal_equal(original, ds)
        finally:
            ds.close()

    def test_with_checksum_roundtrip(self, tmp_path):
        src = _build_dataset(tmp_path / "src.tio")
        mots = tmp_path / "stream.tis"
        rt = tmp_path / "rt.tio"
        file_to_transport(src, mots, use_checksum=True)
        ds = transport_to_file(mots, rt)
        try:
            with SpectralDataset.open(src) as original:
                _assert_signal_equal(original, ds)
        finally:
            ds.close()


# ---------------------------------------------------------- cross-language


def _java_cli_available() -> bool:
    target = REPO_ROOT / "java" / "target" / "classes"
    return target.is_dir() and (target / "global" / "thalion" / "ttio"
                                  / "tools" / "TransportEncodeCli.class").is_file()


def _objc_tool_available(name: str) -> bool:
    return (REPO_ROOT / "objc" / "Tools" / "obj" / name).is_file()


# Java 21 preview-API + FFM flags. The library uses FFM for HDF5 1.14
# VL_BYTES handling; classes compiled with --enable-preview cannot be
# loaded without the same flag at runtime. The java.library.path entry
# is platform-dependent: Linux source-builds land at /usr/local/lib;
# Windows mingw-ucrt64 builds land under $HDF5_HOME/bin. The default
# below covers the Linux CI path; non-Linux environments must set
# TTIO_HDF5_NATIVE so the test loads the right libhdf5_java/{so,dll}.
_HDF5_NATIVE_DEFAULT = "/usr/local/lib"
_HDF5_JAR_DEFAULT = "/usr/local/lib/jarhdf5.jar"


def _hdf5_native_dir() -> str:
    return os.environ.get("TTIO_HDF5_NATIVE", _HDF5_NATIVE_DEFAULT)


def _hdf5_jar_path() -> str:
    return os.environ.get("TTIO_HDF5_JAR", _HDF5_JAR_DEFAULT)


def _java_flags() -> list[str]:
    return [
        "--enable-preview",
        "--enable-native-access=ALL-UNNAMED",
        f"-Djava.library.path={_hdf5_native_dir()}",
    ]


def _build_java_classpath() -> str:
    """Compose the runtime classpath for the Java CLIs.

    Prefers a cached ``target/runtime-classpath.txt`` produced by
    ``mvn dependency:build-classpath`` — that is the project's exact
    transitive dep set, with no risk of pulling in stray jars from a
    shared local Maven cache. Falls back to a deliberately narrow
    glob of the jars actually declared in ``java/pom.xml`` if the
    cache file is absent (e.g. fresh checkout where ``mvn`` has not
    run yet on this machine).
    """
    classes = REPO_ROOT / "java" / "target" / "classes"
    cp_file = REPO_ROOT / "java" / "target" / "runtime-classpath.txt"
    if cp_file.is_file():
        # Maven's resolved classpath already uses os.pathsep on the
        # generating platform — re-emit as a single string with our
        # ``classes`` dir and the HDF5 jar prepended.
        resolved = cp_file.read_text().strip()
        return os.pathsep.join([str(classes), _hdf5_jar_path(), resolved])
    # Conservative fallback: glob only jars whose exact artifact name
    # is in ``java/pom.xml``. Earlier versions of this helper used
    # ``*slf4j*.jar`` which over-pulled ``log4j-slf4j-impl`` from
    # transitive caches and caused ``NoClassDefFoundError: org/apache
    # /log4j/Level`` at HDF5 init time. Keep this list in sync with
    # the runtime-scope deps in ``java/pom.xml``.
    m2 = Path.home() / ".m2" / "repository"
    name_patterns = (
        "Java-WebSocket-*.jar",
        "slf4j-api-2.*.jar",
        "slf4j-simple-2.*.jar",
        "sqlite-jdbc-*.jar",
        "bcprov-jdk18on-*.jar",
        "htsjdk-*.jar",
    )
    jars: list[str] = []
    for pat in name_patterns:
        jars.extend(sorted(str(p) for p in m2.rglob(pat)))
    return os.pathsep.join([str(classes), _hdf5_jar_path(), *jars])


def _run_java(cli: str, *args: str) -> None:
    cp = _build_java_classpath()
    subprocess.run(
        ["java", *_java_flags(), "-cp", cp,
         f"global.thalion.ttio.tools.{cli}", *args],
        check=True, capture_output=True,
    )


def _run_objc(tool: str, *args: str) -> None:
    src_obj = REPO_ROOT / "objc" / "Source" / "obj"
    tool_path = REPO_ROOT / "objc" / "Tools" / "obj" / tool
    env = os.environ.copy()
    env["LD_LIBRARY_PATH"] = (
        f"{src_obj}{os.pathsep}{_hdf5_native_dir()}"
        f"{os.pathsep}{env.get('LD_LIBRARY_PATH', '')}"
    )
    subprocess.run(
        [str(tool_path), *args],
        check=True, capture_output=True, env=env,
    )


@pytest.mark.skipif(
    not _java_cli_available(),
    reason="Java classes not built (run `mvn compile` first)"
)
class TestPythonJavaExchange:

    def test_python_encoded_stream_readable_by_java(self, tmp_path):
        src = _build_dataset(tmp_path / "src.tio")
        mots = tmp_path / "py.tis"
        rt = tmp_path / "rt.tio"
        file_to_transport(src, mots)
        _run_java("TransportDecodeCli", str(mots), str(rt))
        with SpectralDataset.open(src) as a, SpectralDataset.open(rt) as b:
            _assert_signal_equal(a, b)

    def test_java_encoded_stream_readable_by_python(self, tmp_path):
        src = _build_dataset(tmp_path / "src.tio")
        mots = tmp_path / "java.tis"
        rt = tmp_path / "rt.tio"
        _run_java("TransportEncodeCli", str(src), str(mots))
        ds = transport_to_file(mots, rt)
        try:
            with SpectralDataset.open(src) as original:
                _assert_signal_equal(original, ds)
        finally:
            ds.close()


@pytest.mark.skipif(
    not _objc_tool_available("TtioTransportEncode"),
    reason="ObjC tools not built (run `./build.sh` first)"
)
class TestPythonObjCExchange:

    def test_python_encoded_stream_readable_by_objc(self, tmp_path):
        src = _build_dataset(tmp_path / "src.tio")
        mots = tmp_path / "py.tis"
        rt = tmp_path / "rt.tio"
        file_to_transport(src, mots)
        _run_objc("TtioTransportDecode", str(mots), str(rt))
        with SpectralDataset.open(src) as a, SpectralDataset.open(rt) as b:
            _assert_signal_equal(a, b)

    def test_objc_encoded_stream_readable_by_python(self, tmp_path):
        src = _build_dataset(tmp_path / "src.tio")
        mots = tmp_path / "objc.tis"
        rt = tmp_path / "rt.tio"
        _run_objc("TtioTransportEncode", str(src), str(mots))
        ds = transport_to_file(mots, rt)
        try:
            with SpectralDataset.open(src) as original:
                _assert_signal_equal(original, ds)
        finally:
            ds.close()


@pytest.mark.skipif(
    not (_java_cli_available()
         and _objc_tool_available("TtioTransportEncode")),
    reason="Both Java and ObjC tools required"
)
class TestJavaObjCExchange:

    def test_java_encoded_stream_readable_by_objc(self, tmp_path):
        src = _build_dataset(tmp_path / "src.tio")
        mots = tmp_path / "java.tis"
        rt = tmp_path / "rt.tio"
        _run_java("TransportEncodeCli", str(src), str(mots))
        _run_objc("TtioTransportDecode", str(mots), str(rt))
        with SpectralDataset.open(src) as a, SpectralDataset.open(rt) as b:
            _assert_signal_equal(a, b)

    def test_objc_encoded_stream_readable_by_java(self, tmp_path):
        src = _build_dataset(tmp_path / "src.tio")
        mots = tmp_path / "objc.tis"
        rt = tmp_path / "rt.tio"
        _run_objc("TtioTransportEncode", str(src), str(mots))
        _run_java("TransportDecodeCli", str(mots), str(rt))
        with SpectralDataset.open(src) as a, SpectralDataset.open(rt) as b:
            _assert_signal_equal(a, b)
